app/uk/gov/hmrc/digitalservicestax/controllers/RegistrationsController.scala

```
@@ -46,8 +53,51 @@ class RegistrationsController @Inject()(
 
   implicit val ec: ExecutionContext = cc.executionContext
 
-  def submitRegistration(): Action[AnyContent] = Action.async { implicit request =>
-    Future.successful(???) // TODO
+  private def getSafeId(data: Registration)(implicit hc:HeaderCarrier) = {
+    rosmConnector.retrieveROSMDetailsWithoutID(
+      RosmRegisterWithoutIDRequest(
+        false,
+        false,
+        data.company,
+        data.contact
+      )
+    ).map {
+      case Some(r) => r.safeId
+      case _ => ??? // TODO throw some kind of exception
+    }
+  }
+
+
+  import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals.allEnrolments
+
+  private def getUtr(enrolments: Enrolments): Option[String] = {
+    enrolments.getEnrolment("IR-CT").orElse(enrolments.getEnrolment("IR-SA")).flatMap(_.getIdentifier("UTR").map(_.value))
+  }
+
+  def submitRegistration(): Action[JsValue] = Action.async(parse.json) { implicit request =>
+    authorised(AuthProviders(GovernmentGateway)).retrieve(allEnrolments) { enrolments =>
+      withJsonBody[Registration](data => {
+        ((data.utr, getUtr(enrolments), data.useSafeId) match {
```

I think you might have a non-exhaustive match here `(None,None, false)`. Probably doesn't matter for now, or alternatively we could throw an exception.